### PR TITLE
compute(): change default of 'return_type' parameter to 'core'.

### PR DESCRIPTION
**Pull request by skrah**
Implementation for #1551.

### CODE REVIEW

**Comment by skrah** [comment]
@kwmsmith I think this is ready for review.  I had to tweak ```kwargs``` in two places, which is a bit ugly, but also happens elsewhere.

File: blaze/compute/core.py [hunk 1]
```diff
@@ -367,20 +367,20 @@ def compute(expr, d, return_type=no_default, **kwargs):
...
+        returning.
 
     Examples
     --------
     >>> t = symbol('t', 'var * {name: string, balance: int}')
     >>> deadbeats = t[t['balance'] < 0]['name']
 
     >>> data = [['Alice', 100], ['Bob', -50], ['Charlie', -20]]
-    >>> list(compute(deadbeats, {t: data}))
+    >>> list(compute(deadbeats, {t: data}, return_type='native'))
```
Comment (kwmsmith) [inline blaze/compute/core.py]: is `return_type = 'native'` necessary here?  I would have expected `return_type = 'core'` to return a `list`.

Comment (skrah) [inline blaze/compute/core.py]: @kwmsmith Currently the return type is necessary:

```python
>>> list(compute(deadbeats, {t: data}))
Traceback (most recent call last):
...
AttributeError: 'NoneType' object has no attribute 'measure'
```

In many places ```return_type='native'``` is hardcoded (for lack of a better word)
when ```compute``` is used internally.  Example:

https://github.com/blaze/blaze/blob/master/blaze/compute/sql.py#L125

Should that always be the case?  If so, perhaps the argument is missing in 
some places.

Comment (skrah) [inline blaze/compute/core.py]: @kwmsmith Currently the return type is necessary:

```python
>>> list(compute(deadbeats, {t: data}))
Traceback (most recent call last):
...
AttributeError: 'NoneType' object has no attribute 'measure'
```

In many places ```return_type='native'``` is hardcoded (for lack of a better word)
when ```compute``` is used internally.  Example:

https://github.com/blaze/blaze/blob/master/blaze/compute/sql.py#L125

Should that always be the case?  If so, perhaps the argument is missing in 
some places.

**Comment by kwmsmith** [comment]
There are a number of spots when computing against the `python` backend where `compute(..., return_type='native')` is specified and there's an explicit `list()` coercion applied to the result.  The `list()` call shouldn't be necessary, unless there's something I'm misunderstanding.  Can you address those?

File: blaze/compute/core.py (hunk 1 above)
Comment (skrah) [inline blaze/compute/core.py]: Just for reference, the commit that added the arg was 185cc7be7.

Comment (kwmsmith) [inline blaze/compute/core.py]: I see -- `compute(deadbeats, {t: data})` triggers an odo conversion from `itertools.chain` to a DataFrame, and that path appears to be broken in odo (separate issue).

One reason for the `return_type` kwarg was to reduce the amount of explicit `odo`ing and type conversions when calling `bz.compute()`.  To make it more streamlined, please convert

```
list(compute(deadbeats, {t: data}, return_type='native')
```

to

```
compute(deadbeats, {t: data}, return_type=list)
```

**Comment by skrah** [comment]
This works: ```compute(t['name'], data, return_type=list) == [x[0] for x in data]```.

Omitting ```return_type``` entirely doesn't work and ```return_type='native'``` needs the ```list()```.


So I guess I have to figure out the exact return types.

**Comment by kwmsmith** [comment]
> So I guess I have to figure out the exact return types.

Not everywhere -- only in the places where there's a need for an explicit coercion that isn't already a 'native' or 'core' type, or where the `odo` conversion logic needs help.  So the SQL backends can use `return_type='native'`, while the `list` docstring examples should use:

```
compute(expr, data, return_type=list)
```

**Comment by skrah** [comment]
Okay, great.  That saves some work.

**Comment by kwmsmith** [comment]
related: https://github.com/blaze/odo/issues/465

**Comment by postelrich** [comment]
Closing to cleanup old PRs. Feel free to submit a new PR.

**Pull request closed without merging**